File: src/chan/http/FirewallResolver.java
package chan.http;

import android.net.Uri;
import chan.annotation.Extendable;
import chan.annotation.Public;
import chan.content.Chan;
import chan.content.ChanConfiguration;
import com.mishiranu.dashchan.content.net.firewall.FirewallResolvers;
import java.util.Map;
import java.util.Objects;

@Extendable
public abstract class FirewallResolver {
	public static class CheckResult {
		public final boolean resolved;
		public final boolean retransmitOnSuccess;

		public CheckResult(boolean resolved, boolean retransmitOnSuccess) {
			this.resolved = resolved;
			this.retransmitOnSuccess = retransmitOnSuccess;
		}
	}

	public static abstract class Implementation {
		@SuppressWarnings("StaticInitializerReferencesSubClass")
		private static final Implementation INSTANCE = new FirewallResolvers();

		public static Implementation getInstance() {
			return INSTANCE;
		}

		public abstract CheckResult checkResponse(Chan chan, Uri uri, HttpHolder holder, HttpResponse response,
				Identifier identifier, boolean resolve) throws HttpException, InterruptedException;

		public abstract CookieBuilder collectCookies(Chan chan, Uri uri, Identifier identifier, boolean safe);
	}

	@Public
	public static final class Identifier {
		@Public
		public enum Flag {
			@Public USER_AGENT,
			HOST
		}

		@Public public final String userAgent;
		@Public public final boolean defaultUserAgent;
		public final String host;

		public Identifier(String userAgent, boolean defaultUserAgent, String host) {
			this.userAgent = userAgent;
			this.defaultUserAgent = defaultUserAgent;
			this.host = host;
		}

		@Override
		public boolean equals(Object o) {
			if (this == o) return true;
			if (o == null || getClass() != o.getClass()) return false;
			Identifier that = (Identifier) o;
			return defaultUserAgent == that.defaultUserAgent && Objects.equals(userAgent, that.userAgent) && Objects.equals(host, that.host);
		}

		@Override
		public int hashCode() {
			return Objects.hash(userAgent, defaultUserAgent, host);
		}
	}

	@Extendable
	public static abstract class WebViewClient<Result> {
		private final String name;
		private volatile Result result;

		@Public
		public WebViewClient(String name) {
			this.name = name;
		}

		public final String getName() {
			return name;
		}

		@Public
		public final void setResult(Result result) {
			this.result = result;
		}

		public final Result getResult() {
			return result;
		}

		@Extendable
		public boolean onPageFinished(Uri uri, Map<String, String> cookies, String title) {
			return true;
		}

		@Extendable
		public boolean onLoad(Uri initialUri, Uri uri) {
			return true;
		}
	}

	@Public
	public interface Session extends HttpRequest.Preset {
		@Public
		Uri getUri();

		@Override
		HttpHolder getHolder();

		Chan getChan();

		@Public
		ChanConfiguration getChanConfiguration();

		@Public
		Identifier getIdentifier();

		@Public
		Exclusive.Key getKey(Identifier.Flag... flags);

		@Public
		boolean isResolveRequest();

		@Public
		<Result> Result resolveWebView(WebViewClient<Result> webViewClient)
				throws CancelException, InterruptedException;
	}

	@Public
	public static final class CancelException extends Exception {}

	@Extendable
	public interface Exclusive {
		Exclusive FAIL = (session, key) -> false;

		@Public
		interface Key {
			@Public
			String formatKey(String value);

			@Public
			String formatTitle(String value);
		}

		@Extendable
		boolean resolve(Session session, Key key) throws CancelException, HttpException, InterruptedException;
	}

	@Public
	public static class CheckResponseResult {
		public final Exclusive.Key key;
		public final Exclusive exclusive;

		public boolean retransmitOnSuccess;

		@Public
		public CheckResponseResult(Exclusive.Key key, Exclusive exclusive) {
			this.key = key;
			this.exclusive = exclusive;
		}

		@Public
		public CheckResponseResult setRetransmitOnSuccess(boolean retransmitOnSuccess) {
			this.retransmitOnSuccess = retransmitOnSuccess;
			return this;
		}
	}

	@Extendable
	public abstract CheckResponseResult checkResponse(Session session, HttpResponse response) throws HttpException;

	@Extendable
	public void collectCookies(Session session, CookieBuilder cookieBuilder) {}
}

File: src/chan/http/FirewallResolver.kt
package chan.http

import android.net.Uri
import chan.annotation.Extendable
import chan.annotation.Public
import chan.content.Chan
import chan.content.ChanConfiguration
import com.mishiranu.dashchan.content.net.firewall.FirewallResolvers
import java.util.Objects
import kotlin.concurrent.Volatile

@Extendable
abstract class FirewallResolver {
    class CheckResult(val resolved: Boolean, val retransmitOnSuccess: Boolean)

    abstract class Implementation {
        @Throws(HttpException::class, InterruptedException::class)
        abstract fun checkResponse(
            chan: Chan,
            uri: Uri,
            holder: HttpHolder,
            response: HttpResponse?,
            identifier: Identifier,
            resolve: Boolean
        ): CheckResult?

        abstract fun collectCookies(
            chan: Chan,
            uri: Uri?,
            identifier: Identifier,
            safe: Boolean
        ): CookieBuilder

        companion object {
            private val INSTANCE: Implementation = FirewallResolvers()

            @JvmStatic
            fun getInstance(): Implementation = INSTANCE
        }
    }

    @Public
    class Identifier(
        @JvmField @field:Public val userAgent: String?,
        @JvmField @field:Public val defaultUserAgent: Boolean,
        @JvmField val host: String?
    ) {
        @Public
        enum class Flag {
            @Public
            USER_AGENT,
            HOST
        }

        override fun equals(o: Any?): Boolean {
            if (this === o) return true
            if (o == null || javaClass != o.javaClass) return false
            val that = o as Identifier
            return defaultUserAgent == that.defaultUserAgent && userAgent == that.userAgent && host == that.host
        }

        override fun hashCode(): Int {
            return Objects.hash(userAgent, defaultUserAgent, host)
        }
    }

    @Extendable
    abstract class WebViewClient<Result> @Public constructor(@JvmField val name: String?) {
        @Volatile
        internal var result: Result? = null

        @Public
        fun setResult(result: Result?) {
            this.result = result
        }

        fun getResult(): Result? {
            return result
        }

        @Extendable
        open fun onPageFinished(
            uri: Uri,
            cookies: Map<String, String>,
            title: String?
        ): Boolean {
            return true
        }

        @Extendable
        open fun onLoad(initialUri: Uri, uri: Uri): Boolean {
            return true
        }
    }

    @Public
    interface Session : HttpRequest.Preset {
        @Public
        fun getUri(): Uri?

        override val holder: HttpHolder?

        val chan: Chan?

        @Public
        fun getChanConfiguration(): ChanConfiguration?

        @Public
        fun getIdentifier(): Identifier?

        @Public
        fun getKey(vararg flags: Identifier.Flag): Exclusive.Key?

        @Public
        fun isResolveRequest(): Boolean

        @Public
        @Throws(CancelException::class, InterruptedException::class)
        fun <Result : Any> resolveWebView(webViewClient: WebViewClient<Result>): Result?
    }

    @Public
    class CancelException : Exception()

    @Extendable
    fun interface Exclusive {
        @Public
        interface Key {
            @Public
            fun formatKey(value: String?): String?

            @Public
            fun formatTitle(value: String?): String?
        }

        @Extendable
        @Throws(CancelException::class, HttpException::class, InterruptedException::class)
        fun resolve(session: Session, key: Key): Boolean

        companion object {
            val FAIL: Exclusive =
                Exclusive { session: Session, key: Key -> false }
        }
    }

    @Public
    class CheckResponseResult @Public constructor(
        @JvmField val key: Exclusive.Key?,
        @JvmField val exclusive: Exclusive?
    ) {
        @JvmField
        var retransmitOnSuccess: Boolean = false

        @Public
        fun setRetransmitOnSuccess(retransmitOnSuccess: Boolean): CheckResponseResult {
            this.retransmitOnSuccess = retransmitOnSuccess
            return this
        }
    }

    @Extendable
    @Throws(HttpException::class)
    abstract fun checkResponse(session: Session, response: HttpResponse): CheckResponseResult?

    @Extendable
    open fun collectCookies(session: Session, cookieBuilder: CookieBuilder) {
    }
}
